File: services/shared/db/models/knowledge.py
import uuid
from datetime import date, datetime

from sqlalchemy import DATE, INTEGER, TEXT, ForeignKey, String, DateTime, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from .base import Base


class KnowledgeCollection(Base):
    __tablename__ = "knowledge_collection"

    knowledge_collection_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    organization_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    collection_code: Mapped[str] = mapped_column(String(50), nullable=False)
    collection_name: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(TEXT)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="active", server_default="active")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    created_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    updated_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)

    documents: Mapped[list["KnowledgeDocument"]] = relationship("KnowledgeDocument", back_populates="collection")


class KnowledgeDocument(Base):
    __tablename__ = "knowledge_document"

    knowledge_document_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    knowledge_collection_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("knowledge_collection.knowledge_collection_id"),
        nullable=False,
    )
    owner_user_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    document_type: Mapped[str | None] = mapped_column(String(50))
    summary_description: Mapped[str | None] = mapped_column(TEXT)
    version_number: Mapped[str | None] = mapped_column(String(20))
    source_code: Mapped[str | None] = mapped_column(String(100))
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="draft", server_default="draft")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    created_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    updated_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)

    collection: Mapped["KnowledgeCollection"] = relationship("KnowledgeCollection", back_populates="documents")
    governance: Mapped["KnowledgeDocumentGovernance | None"] = relationship(
        "KnowledgeDocumentGovernance", back_populates="document", uselist=False
    )
    tags: Mapped[list["KnowledgeDocumentTag"]] = relationship("KnowledgeDocumentTag", back_populates="document")
    access_entries: Mapped[list["KnowledgeDocumentAccess"]] = relationship(
        "KnowledgeDocumentAccess", back_populates="document"
    )
    chunks: Mapped[list["DocumentChunk"]] = relationship("DocumentChunk", back_populates="document")
    ingestion_jobs: Mapped[list["DocumentIngestionJob"]] = relationship(
        "DocumentIngestionJob", back_populates="document"
    )


class KnowledgeDocumentGovernance(Base):
    __tablename__ = "knowledge_document_governance"

    knowledge_document_governance_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    knowledge_document_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("knowledge_document.knowledge_document_id"),
        unique=True,
        nullable=False,
    )
    classification_level: Mapped[str] = mapped_column(String(50), nullable=False)
    department: Mapped[str | None] = mapped_column(String(255))
    document_owner: Mapped[str | None] = mapped_column(String(255))
    effective_date: Mapped[date | None] = mapped_column(DATE)
    review_date: Mapped[date | None] = mapped_column(DATE)
    expiry_date: Mapped[date | None] = mapped_column(DATE)
    review_status: Mapped[str | None] = mapped_column(TEXT)

    document: Mapped["KnowledgeDocument"] = relationship("KnowledgeDocument", back_populates="governance")


class KnowledgeDocumentTag(Base):
    __tablename__ = "knowledge_document_tag"

    knowledge_document_tag_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    knowledge_document_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("knowledge_document.knowledge_document_id"),
        nullable=False,
    )
    tag_name: Mapped[str] = mapped_column(String(100), nullable=False)
    tag_type: Mapped[str] = mapped_column(String(50), nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="active", server_default="active")

    document: Mapped["KnowledgeDocument"] = relationship("KnowledgeDocument", back_populates="tags")


class KnowledgeDocumentAccess(Base):
    __tablename__ = "knowledge_document_access"

    knowledge_document_access_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    knowledge_document_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("knowledge_document.knowledge_document_id"),
        nullable=False,
    )
    role_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)  # NO ForeignKey
    user_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)  # NO ForeignKey
    access_type: Mapped[str] = mapped_column(String(20), nullable=True, default="read", server_default="read")
    created_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), server_default=func.now())
    created_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)

    document: Mapped["KnowledgeDocument"] = relationship("KnowledgeDocument", back_populates="access_entries")


class DocumentChunk(Base):
    __tablename__ = "document_chunk"

    document_chunk_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    knowledge_document_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("knowledge_document.knowledge_document_id"),
        nullable=False,
    )
    chunk_no: Mapped[int] = mapped_column(INTEGER, nullable=False)
    chunk_title: Mapped[str | None] = mapped_column(String(500))
    chunk_text: Mapped[str | None] = mapped_column(TEXT)
    token_count: Mapped[int | None] = mapped_column(INTEGER)
    processing_status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="pending", server_default="pending"
    )
    parent_chunk_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    document: Mapped["KnowledgeDocument"] = relationship("KnowledgeDocument", back_populates="chunks")


class DocumentIngestionJob(Base):
    __tablename__ = "document_ingestion_job"

    document_ingestion_job_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    knowledge_document_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("knowledge_document.knowledge_document_id"),
        nullable=False,
    )
    job_type: Mapped[str] = mapped_column(String(20), nullable=False, default="initial")
    job_status: Mapped[str] = mapped_column(String(20), nullable=False, default="queued", server_default="queued")
    total_chunks: Mapped[int | None] = mapped_column(INTEGER)
    processed_chunks: Mapped[int | None] = mapped_column(INTEGER)
    error_message: Mapped[str | None] = mapped_column(TEXT)
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    document: Mapped["KnowledgeDocument"] = relationship("KnowledgeDocument", back_populates="ingestion_jobs")

File: frontend/src/lib/api/knowledge.ts
import { apiClient } from './client'

// ─── Types ───────────────────────────────────────────────────────────────────

export interface KnowledgeCollection {
  knowledge_collection_id: string
  collection_name: string
  collection_code: string
  description: string | null
  status: 'active' | 'inactive'
  document_count: number
  created_at: string
}

export interface DocumentGovernance {
  classification_level: 'Public' | 'Internal' | 'Confidential' | 'Restricted'
  document_owner: string | null
  effective_date: string | null
  review_date: string | null
  expiry_date: string | null
  review_status: string | null
  allow_external_llm_usage: boolean
  llm_model_id: string | null
}

export interface DocumentTag {
  knowledge_document_tag_id: string
  tag_name: string
  tag_type: 'domain' | 'sdlc' | 'project_type' | 'keyword'
  status: 'active' | 'inactive'
}

export interface KnowledgeDocument {
  knowledge_document_id: string
  knowledge_collection_id: string
  title: string
  document_type: string | null
  summary_description: string | null
  version_number: string | null
  source_code: string | null
  status: 'draft' | 'active' | 'archived' | 'deleted'
  owner_user_id: string | null
  created_at: string
  updated_at: string
}

export interface DocumentAccessEntry {
  knowledge_document_access_id: string
  user_id: string | null
  access_type: string
}

export interface DocumentListItem extends KnowledgeDocument {
  governance: DocumentGovernance | null
  tags: DocumentTag[]
  access_entries?: DocumentAccessEntry[]
}

export interface GovernanceUpsert {
  classification_level: 'Public' | 'Internal' | 'Confidential' | 'Restricted'
  document_owner?: string
  effective_date?: string
  review_date?: string
  expiry_date?: string
  review_status?: string
  allow_external_llm_usage?: boolean
  llm_model_id?: string
}

export interface TagUpsert {
  tag_name: string
  tag_type: 'domain' | 'sdlc' | 'project_type' | 'keyword'
}

export interface AccessEntry {
  user_id: string
  access_type: 'read' | 'write' | 'admin'
}

export interface DocumentCreate {
  title: string
  document_type?: string
  knowledge_collection_id: string
  summary_description?: string
  version_number?: string
  source_code?: string
}

export interface DocumentUpdate {
  title?: string
  document_type?: string
  knowledge_collection_id?: string
  summary_description?: string
  version_number?: string
  source_code?: string
  status?: 'draft' | 'active' | 'archived' | 'deleted'
}

export interface UploadedFile {
  file_id: string
  original_file_name: string
  mime_type: string
  file_size_bytes: number
  upload_status: string
  storage_path: string
  created_at: string
}

export interface IngestionJob {
  document_ingestion_job_id: string
  knowledge_document_id: string
  job_type: 'initial' | 'reindex' | 'update'
  job_status: 'queued' | 'processing' | 'completed' | 'failed'
  total_chunks: number | null
  processed_chunks: number | null
  progress_pct: number
  error_message: string | null
  started_at: string | null
  completed_at: string | null
}

export interface LLMModel {
  llm_model_id: string
  provider_name: string
  model_name: string
  model_code: string
  is_external: boolean
}

export interface KnowledgeUser {
  user_id: string
  username: string
  email: string
}

export interface DocumentFilters {
  search?: string
  document_type?: string
  knowledge_collection_id?: string
  classification_level?: string
  sdlc?: string
  domain?: string
  persona?: string
  status?: string
  skip?: number
  limit?: number
}

// ─── Collections ─────────────────────────────────────────────────────────────

export const fetchCollections = () =>
  apiClient.get<KnowledgeCollection[]>('/api/v1/knowledge/collections').then((r) => r.data)

export const fetchCollection = (id: string) =>
  apiClient.get<KnowledgeCollection>(`/api/v1/knowledge/collections/${id}`).then((r) => r.data)

export const createCollection = (data: { collection_code: string; collection_name: string; description?: string }) =>
  apiClient.post<KnowledgeCollection>('/api/v1/knowledge/collections', data).then((r) => r.data)

// ─── Documents ───────────────────────────────────────────────────────────────

export const fetchDocuments = (filters: DocumentFilters = {}) => {
  const params: Record<string, string | number> = {}
  if (filters.search) params.search = filters.search
  if (filters.document_type) params.document_type = filters.document_type
  if (filters.knowledge_collection_id) params.knowledge_collection_id = filters.knowledge_collection_id
  if (filters.classification_level) params.classification_level = filters.classification_level
  if (filters.sdlc) params.sdlc = filters.sdlc
  if (filters.domain) params.domain = filters.domain
  if (filters.persona) params.persona = filters.persona
  if (filters.status) params.status = filters.status
  if (filters.skip !== undefined) params.skip = filters.skip
  if (filters.limit !== undefined) params.limit = filters.limit
  return apiClient.get<DocumentListItem[]>('/api/v1/knowledge/documents', { params }).then((r) => r.data)
}

export const fetchDocument = (id: string) =>
  apiClient.get<DocumentListItem>(`/api/v1/knowledge/documents/${id}`).then((r) => r.data)

export const createDocument = (data: DocumentCreate) =>
  apiClient.post<KnowledgeDocument>('/api/v1/knowledge/documents', data).then((r) => r.data)

export const updateDocument = (id: string, data: DocumentUpdate) =>
  apiClient.put<KnowledgeDocument>(`/api/v1/knowledge/documents/${id}`, data).then((r) => r.data)

export const upsertGovernance = (id: string, data: GovernanceUpsert) =>
  apiClient.put(`/api/v1/knowledge/documents/${id}/governance`, data).then((r) => r.data)

export const updateTags = (id: string, tags: TagUpsert[]) =>
  apiClient.put(`/api/v1/knowledge/documents/${id}/tags`, tags).then((r) => r.data)

export const updateAccess = (id: string, entries: AccessEntry[]) =>
  apiClient.put(`/api/v1/knowledge/documents/${id}/access`, entries).then((r) => r.data)

export const updateDocumentStatus = (id: string, status: string) =>
  apiClient.patch(`/api/v1/knowledge/documents/${id}/status`, { status }).then((r) => r.data)

export const reindexDocument = (id: string) =>
  apiClient.post(`/api/v1/knowledge/documents/${id}/reindex`).then((r) => r.data)

// ─── Files ───────────────────────────────────────────────────────────────────

export const uploadFile = (formData: FormData) =>
  apiClient
    .post<UploadedFile>('/api/v1/files/upload', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })
    .then((r) => r.data)

export const getFileMeta = (fileId: string) =>
  apiClient.get<UploadedFile>(`/api/v1/files/${fileId}`).then((r) => r.data)

export const getFileDownloadUrl = (fileId: string) =>
  apiClient.get<{ download_url: string }>(`/api/v1/files/${fileId}/download`).then((r) => r.data)

// ─── References ──────────────────────────────────────────────────────────────

export const fetchLLMModels = () =>
  apiClient.get<LLMModel[]>('/api/v1/knowledge/llm-models').then((r) => r.data)

export const fetchKnowledgeUsers = () =>
  apiClient.get<KnowledgeUser[]>('/api/v1/knowledge/users').then((r) => r.data)

// ─── Ingestion Jobs ───────────────────────────────────────────────────────────

export const fetchJob = (id: string) =>
  apiClient.get<IngestionJob>(`/api/v1/knowledge/jobs/${id}`).then((r) => r.data)

export const fetchJobs = () =>
  apiClient.get<IngestionJob[]>('/api/v1/knowledge/jobs').then((r) => r.data)
